app/src/main/java/fr/free/nrw/commons/category/CategoriesModel.kt

```diff
@@ -50,7 +50,7 @@ class CategoriesModel @Inject constructor(
                 || item.matches("(.*)needing(.*)".toRegex())
                 || item.matches("(.*)taken on(.*)".toRegex())
                 || item.matches(".*0s.*".toRegex())
-                && !item.matches(".*(200|201)0s.*".toRegex())
+                && !item.matches(".*20[0-2]0s.*".toRegex())
```

Comment:
This change achieves what it intends to. Unfortunately, I had to remove the following part of the code:

```kotlin
item.matches(".*(19|20)\\d{2}.*".toRegex())
                && !item.contains(yearInString)
                && !item.contains(prevYearInString)
```

.. to see the effect of this change. Basically every category having a year in it is being filtered out by that code. So, until we finalize what we need to do about that specific regex, could we possibly shuffle the code as follows to ensure at least some of the valid `20xxs` categories at least appear in the search results?

```kotlin
        val is20xxsYear = item.matches(".*0s.*".toRegex())

        if (is20xxsYear) {
            //Check if the year in the form of XX(X)0s is relevant, i.e. in the 2000s or 2010s as stated in Issue #1029
            // If not, check if it is an irrelevant category such as Media_needing_categories_as_of_16_June_2017(Issue #750)
            return !item.matches(".*20[0-2]0s.*".toRegex())
                    || item.matches("(.*)needing(.*)".toRegex())
                    || item.matches("(.*)taken on(.*)".toRegex())
        }
        else {
            // If it is not an year in 20xxs form, then check if item contains a 4-digit word
            // anywhere within the string (.* is wildcard) (Issue #47)
            // And that item does not equal the current year or previous year
            // And if it is an irrelevant category such as Media_needing_categories_as_of_16_June_2017(Issue #750)
            return item.matches(".*(19|20)\\d{2}.*".toRegex())
                    && !item.contains(yearInString)
                    && !item.contains(prevYearInString)
                    || item.matches("(.*)needing(.*)".toRegex())
                    || item.matches("(.*)taken on(.*)".toRegex())
        }
```

Reply:
_(Sorry for the delay, this seems to have skipped my attention while I was dealing with backlog...)_

I've changed that variable from `is20xxsYear` to `mentionsDecade` in your suggested code, as that regex will match e.g. `1920s` and `80s` too, so the original `is20xxsYear` would be confusing name for it.  I've also extracted duplicated checks to `spammyCategory` for easier understanding and maintenance, and made return conditions significantly simpler to read.

I've pushed  simplified and more readable (I hope) version of your code, which now looks like:
```
val mentionsDecade = item.matches(".*0s.*".toRegex())
val recentDecade = item.matches(".*20[0-2]0s.*".toRegex())
val spammyCategory = item.matches("(.*)needing(.*)".toRegex())
                  || item.matches("(.*)taken on(.*)".toRegex())

// always skip irrelevant categories such as Media_needing_categories_as_of_16_June_2017(Issue #750)
if (spammyCategory) {
    return true
}

if (mentionsDecade) {
    // Check if the year in the form of XX(X)0s is recent/relevant, i.e. in the 2000s or 2010s/2020s as stated in Issue #1029
    // Example: "2020s" is OK, but "1920s" is not (and should be skipped)
    return !recentDecade
} else {
    // If it is not an year in decade form (e.g. 19xxs/20xxs), then check if item contains a 4-digit year
    // anywhere within the string (.* is wildcard) (Issue #47)
    // And that item does not equal the current year or previous year
    return item.matches(".*(19|20)\\d{2}.*".toRegex())
            && !item.contains(curYearInString)
            && !item.contains(prevYearInString)
}
```

Does it look more readable/maintainable to you too?

Of course, this code still deals just with fixing the current app logic; after it is merged it would need other changes you mention in https://github.com/commons-app/apps-android-commons/issues/5749#issuecomment-2227279839 to fix things like `Olea (ship, 1982)`

---

I'd also recommend changing that `fun containsYear` to `fun isSpammyCategory` (or something similar); which would IMHO much more precise about what it actually does, and that name is also not factually incorrect (`containsYear` would return `false` even if category did contain year in some cases, like `2024`, which is confusing). 

But that touches too many files, so I'd leave it to someone more experienced if they think it is good idea.

Reply:
> Does it look more readable/maintainable to you too?

Awesome, @mnalis ! This is certainly a lot more readable. Thanks!

> I'd also recommend changing that fun containsYear to fun isSpammyCategory (or something similar); which would IMHO much more precise about what it actually does, and that name is also not factually incorrect (containsYear would return false even if category did contain year in some cases, like 2024, which is confusing).

Precisely. I was about to suggest the same too.

> But that touches too many files, so I'd leave it to someone more experienced if they think it is good idea.

It's easier if you do it by using the refactoring option available in Android Studio. It's not trivial to do it manually. If you're using Android Studio, check out the shortcut to achieve the refactoring [here](https://developer.android.com/studio/intro/keyboard-shortcuts). Otherwise, feel free to let me know and I can do it for you 🙂

Reply:
@sivaraam As I don't use Android Studio, I'd prefer if you would do it, thanks!

Reply:
Done @mnalis. I've done the change. Feel free to give it a cross-check if you wish 🙂

Reply:
Looks good to me @sivaraam ; but should `public boolean containsYear` in `app/src/main/java/fr/free/nrw/commons/repository/UploadRepository.java` which is a wrapper around it (as well as its invocations/tests) **_also_** be renamed?

currently it says:
```
    public boolean containsYear(String name) {
        return categoriesModel.isSpammyCategory(name);
    }
```

Reply:
Makes sense. I seem to have overlooked it. Changed now 👍🏼